src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +23,534 @@ namespace System
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int length = ClampEnd(span, 0, trimElement);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Delimits all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampStart<T>(in ReadOnlySpan<T> span, in T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = 0;
+
+            if (trimElement == null)
+            {
+                for (; start < span.Length; start++)
+                {
+                    if (span[start] != null)
+                        break;
+                }
+            }
+            else
+            {
+                for (; start < span.Length; start++)
+                {
+                    if (!trimElement.Equals(span[start]))
+                        break;
+                }
+            }
+
+            return start;
+        }
+
+        /// <summary>
+        /// Delimits all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="start">The start index from which to being searching.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampEnd<T>(in ReadOnlySpan<T> span, int start, in T trimElement)
+            where T : IEquatable<T>
+        {
+            Debug.Assert(start == 0 || (uint)start < span.Length);
+
+            int end = span.Length - 1;
+
+            if (trimElement == null)
+            {
+                for (; end >= start; end--)
+                {
+                    if (span[end] != null)
+                        break;
+                }
+            }
+            else
+            {
+                for (; end >= start; end--)
+                {
+                    if (!trimElement.Equals(span[start]))
+                        break;
+                }
+            }
+
+            return end - start + 1;
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a set of elements specified
+        /// in a readonly span from the memory.
+        /// </summary>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the memory is returned unaltered.</remarks>
+        public static Memory<T> Trim<T>(this Memory<T> memory, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            switch (trimElements.Length)
```

Comment:
Has the `switch` any advantages over `if - else if`?
When I remember right, the JIT won't inline methods with `switch`.

Reply:
I wasn't aware of that. I had previously used the following which I am happy to change back to if that's a concern:
```csharp
            // Nested if optimizes for N > 1
            if (trimElements.Length <= 1) 
            {
                return trimElements.Length == 0 ? memory : Trim(memory, trimElements[0]);
            }
```
or, more simply, the straightforward conversion to `if (len=0) else if (len=1)`

Reply:
I micro-benched it across a range of lengths of inputs, and it ends up that the _nested if_ (which optimizes for `trimElements.Length > 1`) wins by a _tiny_ margin:
```csharp
|            Method |      Mean |     Error |    StdDev | Ratio |
|------------------ |----------:|----------:|----------:|------:|
|   SingleViaSwitch | 0.0006 ns | 0.0000 ns | 0.0000 ns |  1.00 |
|       SingleViaIf | 0.0006 ns | 0.0000 ns | 0.0000 ns |  1.03 |
| SingleViaNestedIf | 0.0006 ns | 0.0000 ns | 0.0000 ns |  0.99 |
```

Reply:
It's statistically insignificant, but changed to nested-if since it makes intent clearer regardless.